scipy/optimize/_dual_annealing.py

```diff
@@ -0,0 +1,631 @@
+        x = sigmax * self.rand_state.normal()
+        y = self.rand_state.normal()
+        den = np.exp(
+            (self.visiting_param - 1.0) * np.log((np.fabs(y))) / (
+                3.0 - self.visiting_param))
+        return x / den
+
+
+class EnergyState(object):
+    """
+    Class used to record the energy state. At any time, it knows what is the
+    currently used coordinates and the most recent best location.
+
+    Parameters
+    ----------
+    lower : array_like
+        A 1-D numpy ndarray containing lower bounds for generating an initial
+        random components in the `reset` method.
+    upper : array_like
+        A 1-D numpy ndarray containing upper bounds for generating an initial
+        random components in the `reset` method
+        components. Neither NaN or inf are allowed.
+    """
+    # Maximimum number of trials for generating a valid starting point
+    MAX_REINIT_COUNT = 1000
+
+    def __init__(self, lower, upper):
+        self.ebest = None
+        self.current_energy = None
+        self.current_location = None
+        self.xbest = None
+        self.lower = lower
+        self.upper = upper
+
+    def reset(self, func_wrapper, rand_state, x0=None):
+        """
+        Initialize current location is the search domain. If `x0` is not
+        provided, a random location within the bounds is generated.
+        """
+        if x0 is None:
+            self.current_location = self.lower + rand_state.random_sample(
+                len(self.lower)) * (self.upper - self.lower)
+        else:
+            self.current_location = np.copy(x0)
+        init_error = True
+        reinit_counter = 0
+        while init_error:
+            self.current_energy = func_wrapper.fun(self.current_location)
+            if self.current_energy is None:
+                raise ValueError('Objective function is returning None')
+            if (not np.isfinite(self.current_energy) or np.isnan(
+                    self.current_energy)):
+                if reinit_counter >= EnergyState.MAX_REINIT_COUNT:
+                    init_error = False
+                    message = (
+                        'Stopping algorithm because function '
+                        'create NaN or (+/-) infinity values even with '
+                        'trying new random parameters'
+                    )
+                    raise ValueError(message)
+                self.current_location = self.lower + rand_state.random_sample(
+                    self.lower.size) * (self.upper - self.lower)
+                reinit_counter += 1
+            else:
+                init_error = False
+            # If first time reset, initialize ebest and xbest
+            if self.ebest is None and self.xbest is None:
+                self.ebest = self.current_energy
+                self.xbest = np.copy(self.current_location)
+            # Otherwise, we keep them in case of reannealing reset
+
+
+class MarkovChain(object):
```

Comment:
This class is doing more than just a MarkovChain.  The single responsibility principle would argue that, if possible you should split out the part that's doing a single MarkovChain from the rest of what this class is doing.  If that is not possible or appropriate, then you should probably rename this class

Reply:
Classed renamed into StrategyChain which reflects better the purpose.